File: 0033-search-in-rotated-sorted-array/solution.cpp
class Solution {
public:
    int search(vector<int>& arr, int target) {
       int low = 0;
       int high = arr.size()-1;
       while (low <= high) {
          int mid = (low + high)   / 2;
          if (arr[mid] == target) return mid;
          
           if (arr[mid] >= arr[low]) {
              if (arr[mid] > target and arr[low] <= target) {
                  high = mid-1;
              } else {
                  low = mid+1;
              }
          } else {
              if (arr[high] >= target and arr[mid] < target) {
                  low = mid+1;
              } else {
                  high = mid-1;
              }
          }
        }
        return -1;
    }
};

File: 0033-search-in-rotated-sorted-array/solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        h = len(nums) - 1
        while l < h:
            mid = (l + h) // 2
            if (nums[l] <= nums[mid] and nums[l] <= target <= nums[mid]) or (nums[mid] <= nums[h] and (target <= nums[mid] or target > nums[h])):
                h = mid
            else:
                l = mid + 1
        return -1 if nums[l] != target else l
